### ISSUE TITLE
Error when compiling dev-haskell/cabal-syntax

### ISSUE THREAD

**Issue by sayurc** [gentoo-haskell/gentoo-haskell#1570]
I updated from GHC 9.6.6 to GHC 9.8.2 but when I run haskell-updater I get an error saying that the package Alex could not be found. Shouldn’t it have been installed automatically?

The output of `emerge --info '=dev-haskell/cabal-syntax-3.12.0.0::haskell'`:
```
Portage 3.0.66.1 (python 3.12.8-final-0, default/linux/amd64/23.0/desktop/plasma, gcc-14, glibc-2.40-r5, 6.6.62-gentoo-dist x86_64)
=================================================================
                         System Settings
=================================================================
System uname: Linux-6.6.62-gentoo-dist-x86_64-AMD_Ryzen_5_5500U_with_Radeon_Graphics-with-glibc2.40
KiB Mem:     6906428 total,    704148 free
KiB Swap:   16777212 total,  15926748 free
Timestamp of repository gentoo: Mon, 30 Dec 2024 01:18:23 +0000
Head commit of repository gentoo: 0542d079a25b931102a660b5c5ba0cccdf4a1e55

Timestamp of repository guru: Sun, 29 Dec 2024 18:03:14 +0000
Head commit of repository guru: 76952bd4abea2a7f305b9012f9a8599fcb6deee1

Timestamp of repository haskell: Mon, 30 Dec 2024 04:33:28 +0000
Head commit of repository haskell: 5a3416d35b673f3f7bec51ebbc1e9113cd23b538

sh bash 5.2_p37
ld GNU ld (Gentoo 2.43 p3) 2.43.1
ccache version 4.10.2 [disabled]
app-misc/pax-utils:        1.3.8::gentoo
app-shells/bash:           5.2_p37::gentoo
dev-build/autoconf:        2.13-r8::gentoo, 2.72-r1::gentoo
dev-build/automake:        1.16.5-r2::gentoo
dev-build/cmake:           3.30.6::gentoo
dev-build/libtool:         2.4.7-r4::gentoo
dev-build/make:            4.4.1-r100::gentoo
dev-build/meson:           1.5.2::gentoo
dev-lang/perl:             5.40.0::gentoo
dev-lang/python:           3.12.8::gentoo, 3.13.1::gentoo
dev-lang/rust-bin:         1.82.0-r101::gentoo
dev-util/ccache:           4.10.2-r1::gentoo
llvm-core/clang:           19.1.4::gentoo
llvm-core/lld:             19.1.4::gentoo
llvm-core/llvm:            18.1.8-r6::gentoo, 19.1.4::gentoo
sys-apps/baselayout:       2.17::gentoo
sys-apps/openrc:           0.55.1::gentoo
sys-apps/sandbox:          2.39::gentoo
sys-devel/binutils:        2.43-r2::gentoo
sys-devel/binutils-config: 5.5.2::gentoo
sys-devel/gcc:             14.2.1_p20241116::gentoo
sys-devel/gcc-config:      2.11::gentoo
sys-kernel/linux-headers:  6.6-r1::gentoo (virtual/os-headers)
sys-libs/glibc:            2.40-r5::gentoo
Repositories:

gentoo
    location: /var/db/repos/gentoo
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/gentoo.git
    priority: -1000
    volatile: False

guru
    location: /var/db/repos/guru
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/guru.git
    masters: gentoo
    volatile: False

haskell
    location: /var/db/repos/haskell
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/haskell.git
    masters: gentoo
    volatile: False

Binary Repositories:

gentoobinhost
    priority: 1
    sync-uri: https://distfiles.gentoo.org/releases/amd64/binpackages/23.0/x86-64

ACCEPT_KEYWORDS="amd64"
ACCEPT_LICENSE="*"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-march=native -mtune=native -O2"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /usr/share/config /usr/share/gnupg/qualified.txt"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/revdep-rebuild /etc/sandbox.d"
CXXFLAGS="-march=native -mtune=native -O2"
DISTDIR="/var/cache/distfiles"
ENV_UNSET="CARGO_HOME DBUS_SESSION_BUS_ADDRESS DISPLAY GDK_PIXBUF_MODULE_FILE GOBIN GOPATH PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR XDG_STATE_HOME"
FCFLAGS="-march=native -mtune=native -O2"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs binpkg-multi-instance buildpkg-live clean-logs config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync merge-wait multilib-strict network-sandbox news parallel-fetch pid-sandbox pkgdir-index-trusted preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
FFLAGS="-march=native -mtune=native -O2"
GENTOO_MIRRORS="https://gentoo.c3sl.ufpr.br/"
LANG="en_US.utf8"
LDFLAGS="-Wl,-O1 -Wl,--as-needed -Wl,-z,pack-relative-relocs"
LEX="flex"
MAKEOPTS="-j4"
PKGDIR="/var/cache/binpkgs"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/var/tmp"
SHELL="/usr/bin/zsh"
USE="X a52 aac acl acpi activities alsa amd64 bluetooth branding bzip2 cairo cdda cdr cet crypt cups dbus declarative dri dts dvd dvdr elogind encode exif flac gdbm gif gpm gtk gui iconv icu ipv6 jpeg kde kf6compat kwallet lcms libnotify libtirpc mad mng mp3 mp4 mpeg multilib ncurses networkmanager nls ogg opengl openmp pam pango pcre pdf pipewire plasma png policykit ppds pulseaudio qml qt5 qt6 readline screencast sdl seccomp secureboot semantic-desktop sound spell ssl startup-notification svg test-rust tiff tkip truetype udev udisks unicode upower usb vorbis vulkan wayland widgets wxwidgets x264 xattr xcb xft xml xv xvid zlib" ABI_X86="64" ADA_TARGET="gcc_13" APACHE2_MODULES="authn_core authz_core socache_shmcb unixd actions alias auth_basic authn_anon authn_dbm authn_file authz_dbm authz_groupfile authz_host authz_owner authz_user autoindex cache cgi cgid dav dav_fs dav_lock deflate dir env expires ext_filter file_cache filter headers include info log_config logio mime mime_magic negotiation rewrite setenvif speling status unique_id userdir usertrack vhost_alias" CALLIGRA_FEATURES="karbon sheets words" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="mmx mmxext sse sse2 aes avx avx2 f16c fma3 pclmul popcnt rdrand sha sse3 sse4_1 sse4_2 sse4a ssse3" ELIBC="glibc" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax navcom oceanserver oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 tsip tripmate tnt ublox" GUILE_SINGLE_TARGET="3-0" GUILE_TARGETS="3-0" INPUT_DEVICES="libinput" KERNEL="linux" LCD_DEVICES="bayrad cfontz glk hd44780 lb216 lcdm001 mtxorb text" LUA_SINGLE_TARGET="lua5-1" LUA_TARGETS="lua5-1" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php8-2" POSTGRES_TARGETS="postgres16" PYTHON_SINGLE_TARGET="python3_12" PYTHON_TARGETS="python3_12" RUBY_TARGETS="ruby32" VIDEO_CARDS="amdgpu radeonsi" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipp2p iface geoip fuzzy condition tarpit sysrq proto logmark ipmark dhcpmac delude chaos account"
Unset:  ADDR2LINE, AR, ARFLAGS, AS, ASFLAGS, CC, CCLD, CONFIG_SHELL, CPP, CPPFLAGS, CTARGET, CXX, CXXFILT, ELFEDIT, EMERGE_DEFAULT_OPTS, EXTRA_ECONF, F77FLAGS, FC, GCOV, GPROF, INSTALL_MASK, LC_ALL, LD, LFLAGS, LIBTOOL, LINGUAS, MAKE, MAKEFLAGS, NM, OBJCOPY, OBJDUMP, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS, PYTHONPATH, RANLIB, READELF, RUSTFLAGS, SIZE, STRINGS, STRIP, YACC, YFLAGS
```
The build log:
```
 * Package:    dev-haskell/cabal-syntax-3.12.0.0:0/3.12.0.0
 * Repository: haskell
 * Maintainer: haskell@gentoo.org
 * USE:        abi_x86_64 amd64 elibc_glibc kernel_linux
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
>>> Unpacking source...
>>> Unpacking cabal-syntax-3.12.0.0.tar.gz to /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work
=== Skipping unpack of cabal-syntax-3.12.0.0-rev1.cabal
>>> Source unpacked in /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work
>>> Preparing source in /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0 ...
 * Using revised .cabal file from Hackage: revision 1
dos2unix: converting file /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0/Cabal-syntax.cabal to Unix format...
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0 ...
 * GHC version: 9.8.2 PM:9.8.2-r2
 * Using cabal-3.10.3.0.
/usr/bin/ghc -hide-all-packages -package Cabal -package array -package base -package binary -package bytestring -package containers -package deepseq -package directory -package exceptions -package filepath -package haskeline -package mtl -package parsec -package pretty -package process -package stm -package template-haskell -package terminfo -package text -package time -package transformers -package unix -package xhtml --make /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0/Setup.hs -j4 +RTS -A256M -qb0 -RTS -threaded -dynamic -o setup
[1 of 2] Compiling Main             ( /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0/Setup.hs, /var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0/Setup.o )
[2 of 2] Linking setup
./setup configure --ghc --prefix=/usr --with-compiler=/usr/bin/ghc --with-hc-pkg=/usr/bin/ghc-pkg --prefix=/usr --libdir=/usr/lib64 --libsubdir=cabal-syntax-3.12.0.0/ghc-9.8.2 --datadir=/usr/share/ --datasubdir=cabal-syntax-3.12.0.0/ghc-9.8.2 --ghc-options=-j4 +RTS -A256M -qb0 -RTS --with-ar=x86_64-pc-linux-gnu-ar --ghc-option=-optc-march=native --ghc-option=-optc-mtune=native --ghc-option=-optc-O2 --ghc-option=-optl-Wl,-O1 --ghc-option=-optl-Wl,--as-needed --ghc-option=-optl-Wl,-z,pack-relative-relocs --disable-executable-stripping --docdir=/usr/share/doc/cabal-syntax-3.12.0.0 --verbose --enable-shared --enable-executable-dynamic --sysconfdir=/etc --disable-library-stripping
Using Parsec parser
Configuring Cabal-syntax-3.12.0.0...
Dependency array >=0.4.0.1 && <0.6: using array-0.5.6.0
Dependency base >=4.11.1.0 && <5: using base-4.19.1.0
Dependency binary >=0.7 && <0.9: using binary-0.8.9.1
Dependency bytestring >=0.10.0.0 && <0.13: using bytestring-0.12.1.0
Dependency containers >=0.5.0.0 && <0.8: using containers-0.6.8
Dependency deepseq >=1.3.0.1 && <1.6: using deepseq-1.5.0.0
Dependency directory >=1.2 && <1.4: using directory-1.3.8.5
Dependency filepath >=1.3.0.1 && <1.6: using filepath-1.4.300.1
Dependency mtl >=2.1 && <2.4: using mtl-2.3.1
Dependency parsec >=3.1.13.0 && <3.2: using parsec-3.1.17.0
Dependency pretty >=1.1.1 && <1.2: using pretty-1.1.3.6
Dependency text >=1.2.3.0 && <1.3 || >=2.0 && <2.2: using text-2.1.1
Dependency time >=1.4.0.1 && <1.15: using time-1.12.2
Dependency transformers >=0.3 && <0.4 || >=0.4.1.0 && <0.7: using
transformers-0.6.1.0
CallStack (from HasCallStack):
  withMetadata, called at libraries/Cabal/Cabal/src/Distribution/Simple/Utils.hs:368:14 in Cabal-3.10.3.0-inplace:Distribution.Simple.Utils
Error: setup: The program 'alex' is required but it could not be found.

 * ghc-pkg check: 'checking for other broken packages:'
 * Detected outdated packages: dev-haskell/cabal-syntax:0
 * ERROR: dev-haskell/cabal-syntax-3.12.0.0::haskell failed (configure phase):
 *   //==-- Please, run 'haskell-updater' to fix outdated packages --==//
 * 
 * Call stack:
 *     ebuild.sh, line  136:  Called src_configure
 *   environment, line 1994:  Called haskell-cabal_src_configure
 *   environment, line 1652:  Called cabal-configure
 *   environment, line  579:  Called cabal-show-brokens-and-die 'setup configure failed'
 *   environment, line  726:  Called cabal-show-old
 *   environment, line  732:  Called cabal-die-if-nonempty 'outdated' 'dev-haskell/cabal-syntax:0'
 *   environment, line  600:  Called die
 * The specific snippet of code:
 *       die "//==-- Please, run 'haskell-updater' to fix ${breakage_type} packages --==//"
 * 
 * If you need support, post the output of `emerge --info '=dev-haskell/cabal-syntax-3.12.0.0::haskell'`,
 * the complete build log and the output of `emerge -pqv '=dev-haskell/cabal-syntax-3.12.0.0::haskell'`.
 * The complete build log is located at '/var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/temp/environment'.
 * Working directory: '/var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0'
 * S: '/var/tmp/portage/dev-haskell/cabal-syntax-3.12.0.0/work/Cabal-syntax-3.12.0.0'
```
Output of `emerge -pqv '=dev-haskell/cabal-syntax-3.12.0.0::haskell'`:
```
[ebuild     U ] dev-haskell/cabal-syntax-3.12.0.0 [3.10.3.0] USE="-doc -hoogle -hscolour -profile" 
```

**Comment by hololeap** [comment]
While it's technically possible to upgrade `Cabal-syntax` (along with `Cabal`, `text` and a couple other packages that come bundled with GHC), it is safest to use the version that comes with the GHC version you're using. This generally means using the `package.unmask` file for your GHC version ([`scripts/package.unmask/ghc-9.8`](https://github.com/gentoo-haskell/gentoo-haskell/blob/master/scripts/package.unmask/ghc-9.8) in this case) and removing any others (`ghc-9.10` for instance) so that it sticks with the versions that are known to work.

In other words, stay on `Cabal-syntax-3.10.3.0` for now. If you really need the newer version of Cabal, try getting `alex` installed first before unmasking the newer versions. ghc-9.10 also requires `alex` and I'm not sure what the best way to handle it is (possibly making a statically-linked binpkg, `alex-bin`, or something).

You can leave this bug open since it should technically be possible to upgrade Cabal-syntax past what comes bundled with GHC.

**Comment by sayurc** [comment]
I unmasked every version of GHC so I wasn’t sure which version of Cabal to install to match the GHC version. I’ll unmask only 9.8.2 then.
